### ISSUE TITLE
Shared circuit breaker when registering a resilience handler with ConfigureHttpClientDefaults

### ISSUE THREAD

**Issue by sveinungf** [dotnet/extensions#5021]
### Description

When registering the standard resilience handler using `builder.Services.ConfigureHttpClientDefaults(x => x.AddStandardResilienceHandler)`, all `HttpClient`s seem to be using a shared circuit breaker. When one client is causing the circuit to open, the circuit opens for all clients.

### Reproduction Steps

Here is my project file:
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Http.Resilience" Version="8.2.0" />
  </ItemGroup>

</Project>
```

Here is my appsettings.json (modified to be able to more easily reproduce the issue):
```json
{
  "RetryOptions": {
    "Retry": {
      "Delay": "00:00:00",
      "MaxRetryAttempts": 1
    }
  }
}
```

And here is the Program.cs:
```csharp
var builder = WebApplication.CreateBuilder(args);

var section = builder.Configuration.GetSection("RetryOptions");

var succeedingClientBuilder = builder.Services.AddHttpClient<SucceedingClient>(x =>
    x.BaseAddress = new Uri("https://ipv4.icanhazip.com/"));

var failingClientBuilder = builder.Services.AddHttpClient<FailingClient>(x =>
    x.BaseAddress = new Uri("http://the-internet.herokuapp.com/status_codes/500"));

var useHttpClientDefaults = true;
if (useHttpClientDefaults)
{
    builder.Services.ConfigureHttpClientDefaults(x => x.AddStandardResilienceHandler(section));
}
else
{
    succeedingClientBuilder.AddStandardResilienceHandler(section);
    failingClientBuilder.AddStandardResilienceHandler(section);
}

var app = builder.Build();

app.MapGet("/test", async (SucceedingClient succeedingClient, FailingClient failingClient) =>
{
    var succeedingClientResult = await succeedingClient.Get();
    var failingClientResult = await failingClient.Get();

    return $"""
        SucceedingClient result: {succeedingClientResult}
        FailingClient result: {failingClientResult}
        """;
});

app.Run();

public class SucceedingClient(HttpClient client)
{
    public async Task<string> Get()
    {
        try
        {
            using var response = await client.GetAsync("");
            return $"Status code {response.StatusCode}";
        }
        catch (Exception e)
        {
            return $"Exception {e.Message}";
        }
    }
}

public class FailingClient(HttpClient client)
{
    public async Task<string> Get()
    {
        try
        {
            using var response = await client.GetAsync("");
            return $"Status code {response.StatusCode}";
        }
        catch (Exception e)
        {
            return $"Exception {e.Message}";
        }
    }
}
```

Run this sample code and continuously hit the `/test` endpoint until the circuit breaker kicks in.

### Expected behavior

I would expect the circuit would only open for the client that is facing issues. From the reproduction steps, the circuit should not open for `SucceedingClient` when it opens for `FailingClient`.

### Actual behavior

The circuit breaker opens for both the `SucceedingClient` and the `FailingClient`, even though only `FailingClient` is receiving status code 500.

### Known Workarounds

Use `AddStandardResilienceHandler` on each HttpClient instead of using `ConfigureHttpClientDefaults`.

### Configuration

<details>
<summary>dotnet --info</summary>
.NET SDK:
 Version:           8.0.200
 Commit:            438cab6a9d
 Workload version:  8.0.200-manifests.e575128c

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22631
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.200\

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.2
  Architecture: x64
  Commit:       1381d5ebd2

.NET SDKs installed:
  8.0.200 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.27 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found
</details>

**Comment by geeknoid** [comment]
@martintmk Do you have time to take a look please?

**Comment by iliar-turdushev** [comment]
The issue occurs because `ConfigureHttpClientDefaults(x => x.AddStandardResilienceHandler())` registers a single resilience pipeline with a name `standard` that is then shared by both `SucceedingClient` and `FailingClient`. As a result, they both share the circuit breaker of the `standard` pipeline.

When we use two separate statements to register the resilience pipeline
```csharp
succeedingClientBuilder.AddStandardResilienceHandler(section)
failingClientBuilder.AddStandardResilienceHandler(section)
```
we create two distinct resilience pipelines named `SucceedingClient-standard` and `FailingClient-standard`, and each of them has its own circuit breaker. Each of these pipelines is then used by the respective `HttpClient`s.

The reason for that is how `AddStandardResilienceHandler()` and `ConfigureHttpClientDefaults()` methods work. `AddStandardResilienceHandler()` relies on `IHttpClientBuilder.Name` when registering and resolving the corresponding resilience pipeline. Whenever you call `AddStandardResilienceHandler()` you register a resilience pipeline for the given `IHttpClientBuilder.Name`, and when you create an `HttpClient` the corresponding resilience pipeline is retrieved by the name or type of the `HttpClient`, i.e. by the given `IHttpClientBuilder.Name`. `ConfigureHttpClientDefaults()` executes its lambda `x => x.AddStandardResilienceHandler()` only once, and the `x` argument is an `IHttpClienBuilder` with the `Name=null`, i.e. it doesn't have information about the actual name of `HttpClient`. It doesn't execute the lambda for each registered `HttpClient`. As a result, we have a single resilience pipeline (and hence a single shared circuit breaker) which is applied to all `HttpClient`s regardless of their name or type.

@JamesNK Can `ConfigureHttpClientDefaults(Action<IHttpClientBuilder> configure)` be updated to provide the actual name/type of the `HttpClient` being resolved? As I understand, currently its `configure` action is executed immediately when `ConfigureHttpClientDefaults()` is invoked, therefore it doesn't have the actual `Name` of `HttpClient`. And to be able to provide the `Name` it should be updated to execute the `configure` action when the actual `HttpClient` is being resolved/created, i.e. when there is information about its type/name. Can we update `ConfigureHttpClientDefaults()` in a such way?

### Workaround

@sveinungf As a workaround, you can configure the standard resilience pipeline to be applied per HTTP request authority:

```csharp
builder.Services.ConfigureHttpClientDefaults(x => x
    .AddStandardResilienceHandler(section)
    .SelectPipelineByAuthority());
```

If you do so, you'll have distinct resilience pipelines per each HTTP request authority. It'll solve the shared circuit breaker issue. You can even specify a custom criteria how to select resilience pipelines using the method `SelectPipelineBy`.

**Comment by JamesNK** [comment]
> @JamesNK Can ConfigureHttpClientDefaults(Action<IHttpClientBuilder> configure) be updated to provide the actual name/type of the HttpClient being resolved?

`IHttpClientBuilder` has a name property on it. Have you tried using that value?

This feature lives in dotnet/runtime - https://github.com/dotnet/runtime/blob/55c904024601c133f8ad081bc704c3c1fc5c7c9b/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientFactoryServiceCollectionExtensions.cs#L80. I don't work on the HTTP client team and I don't have time to investigate more. I suggest creating an issue there to talk about it with the HTTP client folks.

**Comment by iliar-turdushev** [comment]
> `IHttpClientBuilder` has a name property on it. Have you tried using that value?

Yes, I have. `IHttpClientBuilder.Name` is always null.

> I don't work on the HTTP client team and I don't have time to investigate more. I suggest creating an issue there to talk about it with the HTTP client folks.

Got it. I noticed that it was you who proposed the API [#87914](https://github.com/dotnet/runtime/issues/87914) and then implemented it [#87953](https://github.com/dotnet/runtime/pull/87953), therefore I mentioned you in the discussion and asked for help. Sure, I'll create an issue and start discussion with HTTP client folks. Thank you.

**Comment by iliar-turdushev** [comment]
Here is an [issue](https://github.com/dotnet/runtime/issues/101719) in dotnet/runtime to discuss `ConfigureHttpClientDefaults`.

**Comment by RussKie** [comment]
If I understood the problem and possible solution options, and if the following works as expected:
```cs
var services = new ServiceCollection();

services.AddHttpClient("SucceedingClient", x => x.BaseAddress = new Uri("https://succeeding.api/"))
    .AddStandardResilienceHandler();
services.AddHttpClient("FailingClient", x => x.BaseAddress = new Uri("https://failing.api/"))
    .AddStandardResilienceHandler();
```

...then is it not a matter of documenting the behaviours and scenarios in which those occur?

Also, what would be the behaviour if the following code was written?
```cs
var builder = WebApplication.CreateBuilder(args);

var section = builder.Configuration.GetSection("RetryOptions");

builder.Services.ConfigureHttpClientDefaults(x => x.AddStandardResilienceHandler(section));

var succeedingClientBuilder = builder.Services.AddHttpClient<SucceedingClient>(x => x.BaseAddress = new Uri("https://ipv4.icanhazip.com/"))4
                                           .AddStandardResilienceHandler();
var failingClientBuilder = builder.Services.AddHttpClient<FailingClient>(x => x.BaseAddress = new Uri("http://the-internet.herokuapp.com/status_codes/500"))
                                           .AddStandardResilienceHandler();
```

**Comment by iliar-turdushev** [comment]
> is it not a matter of documenting the behaviours and scenarios in which those occur?

That's what I think too. In my opinion, we need to document that registering the Standard Resilience handler as default `HttpClient` configuration could lead to "shared circuit breaker" issue, and the issue could be solve either by configuring the standard pipeline to be applied per each HTTP request host or per some custom condition based on HTTP request or by registering the pipeline for each named `HttpClient`.

> Also, what would be the behaviour if the following code was written?

At the moment it would lead to registering Standard Resilience handler more than once, which is not correct, because it is not reasonable to have more than one standard resilience handlers in a single `HttpClient`. Perfectly, the code you provided would lead to replacing Standard Resilience handler registered as default configuration with Standard Resilience handlers registered for those particular `HttpClient`s. But that's not how it works now. And there is a corresponding bug reported https://github.com/dotnet/extensions/issues/4814. If we fix that bug then the code you provided could be another way to fix the problem described in the current issue.

I'd like to keep this issue as a documentation gap, and in the scope of the https://github.com/dotnet/extensions/issues/4814 issue make your sample work correctly. What do you think?

**Comment by RussKie** [comment]
We could also close #4818 as dup of this one, and continue here :)

To sum up, we have two issues:
1. A documentation gap which will address the immediate issue (more or less a tactical fix).
2. A possible enhancement that will allow overriding the HTTP client resilience settings.

**Comment by iliar-turdushev** [comment]
Regarding the enhancement that will allow to override the Standard Resilience Handler.

As @martintmk pointed out we can implement an improvement when a second (and any following) attempt to register the Standard Resilience Handler doesn't add a new handler, but changes the configuration of the first registered handler. If we implement that then the following code sample:

```csharp
var services = new ServiceCollection();
services.ConfigureHttpClientDefaults((IHttpClientBuilder x) => x.AddStandardResilienceHandler());

services.AddHttpClient("FailingClient", x => x.BaseAddress = new Uri("https://failing.api"));

services.AddHttpClient("SucceedingClient", x => x.BaseAddress = new Uri("https://succeeding.api/"))
    AddStandardResilienceHandler();
```

will result in having two `HttpClient`s:

- "FailingClient" uses the default standard resilience handler. It will share the default handler with any other HttpClient that doesn't explicitly register its own resilience handler.
- "SucceedingClient" will have its own resilience handler. Its handler will replace the default handler. Failures caused by the "FailingClient" will not affect the resilience handler of the "SucceedingClient".

And this solves the issue with the "shared circuit breaker". The downside is that a customer has to explicitly register a named client and register the standard resilience for it.

But there still will be issues we'll need to address. They are listed below.

### 1. StandardResilienceHandler vs StandardHedgingHandler

As @martintmk said, having two standard resilience handlers in HttpClient would break it. Similarly, having two standard hedging handlers or combination of hedging and resilience handlers will break HttpClient. With the approach/solution proposed above we will need not only overwrite the configuration of the handler but also its type, e.g. if someone registers standard resilience handler and then standard hedging handler, then he/she will get HttpClient with the only standard hedging handler, i.e. the last registered handler will "win".

### 2. What if we need to change the order of the resilience handler in HttpClient pipeline?

With the proposed approach it is not achievable. We can consider introducing a method `RemoveAllStandardResilienceHandlers` similar to [RemoveAllLoggers](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientbuilderextensions.removeallloggers?view=net-8.0). It will help to remove all standard resilience handlers and put your resilience handler in the required place in the request pipeline. But since there is no such need at the moment we can get back to this when the need occurs.

### 3. Resilience handlers could be registered by 3rd party libraries

If some 3rd party library registers its own resilience handler then with the the last registered resilience handler "wins" approach we'll get HttpClient with the resilience handler provided by 3rd party's library which could be undesirable. Do we need to introduce a special method with `TryAdd*` semantics that will not overwrite the already registered handler? We'll need to address this too, but I think we can skip for now this since there is no such a request from customers.

---

**To conclude,** I suggest to implement the approach suggested above: the last registered resilience or hedging handler overwrites settings of the first registered handler. We don't allow more than one standard resilience or hedging handler in HttpClient. We'll postpone solving potential issues 2 and 3 mentioned above until customers come to us with corresponding asks.

So, what do you think?

**Comment by RussKie** [comment]
> Would it not be more compatible to just have .ReplaceXXX or a .ClearResilienceHandlers() then .AddXXX?
> 
> This would allow us to imperatively do it with full knowledge of what we're doing and why without breaking anything already there.

How would the API and their use look like?

**Comment by JohnGalt1717** [comment]
Something like this:

```
builder.Services.AddHttpClient<SucceedingClient>(x => x.BaseAddress = new Uri("https://ipv4.icanhazip.com/"))
   .ClearResilienceHandlers();
   .AddResilienceHandler(x => {stuff here});
```

and 

```
builder.Services.AddHttpClient<SucceedingClient>(x => x.BaseAddress = new Uri("https://ipv4.icanhazip.com/"))
   .ReplaceStandardResilienceHandler(x => {stuff here});
```

**Comment by iliar-turdushev** [comment]
@JohnGalt1717 Thank you for the feedback. A comment above https://github.com/dotnet/extensions/issues/5021#issuecomment-2359380710 suggests to implement `AddStandardResilienceHandler` method in a way that it will have behavior of a method `AddOrReplaceStandardResilienceHandler`.

Regarding `RemoveAll*` method. I think we might want to introduce such a method in order to enable users to change order of the resilience handler. In the scope of this task I propose only changing behavior of the `AddStandardResilienceHandler` method.

**Comment by iliar-turdushev** [comment]
We're going to close this issue as duplicate of #5695. Here is a [solution](https://github.com/dotnet/extensions/issues/5695#issuecomment-2582138364) we are going to implement.
